File: frontend/lib/api/block-types.ts
/**
 * Block Types API Client
 * 
 * Provides type-safe access to the Block Registry API for fetching
 * available block types and their configurations.
 */

export interface BlockSubBlock {
  id: string;
  type: 'text' | 'textarea' | 'number' | 'dropdown' | 'toggle' | 'slider' | 'multi-select';
  title: string;
  required?: boolean;
  placeholder?: string;
  default?: any;
  min?: number;
  max?: number;
  step?: number;
  options?: Array<{ label: string; value: string }> | string[];
}

export interface BlockTypeConfig {
  type: string;
  name: string;
  description: string;
  category: string;
  bg_color: string;
  icon: string;
  sub_blocks: BlockSubBlock[];
  inputs: Record<string, any>;
  outputs: Record<string, any>;
  auth_mode?: string;
  docs_link?: string;
}

export interface BlockTypesByCategory {
  [category: string]: BlockTypeConfig[];
}

export interface BlockTypeStats {
  total_blocks: number;
  by_category: Record<string, number>;
  block_types: string[];
  categories: string[];
}

class BlockTypesAPI {
  private baseUrl: string;

  constructor() {
    this.baseUrl = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
  }

  /**
   * Get all block types grouped by category
   */
  async getAllBlockTypes(): Promise<BlockTypesByCategory> {
    const response = await fetch(`${this.baseUrl}/api/agent-builder/block-types`);
    if (!response.ok) {
      throw new Error('Failed to fetch block types');
    }
    return response.json();
  }

  /**
   * Get block types for a specific category
   */
  async getBlockTypesByCategory(category: string): Promise<BlockTypeConfig[]> {
    const response = await fetch(
      `${this.baseUrl}/api/agent-builder/block-types/category/${category}`
    );
    if (!response.ok) {
      throw new Error(`Failed to fetch block types for category: ${category}`);
    }
    return response.json();
  }

  /**
   * Get configuration for a specific block type
   */
  async getBlockType(blockType: string): Promise<BlockTypeConfig> {
    const response = await fetch(
      `${this.baseUrl}/api/agent-builder/block-types/${blockType}`
    );
    if (!response.ok) {
      throw new Error(`Failed to fetch block type: ${blockType}`);
    }
    return response.json();
  }

  /**
   * Get input/output schema for a specific block type
   */
  async getBlockTypeSchema(blockType: string): Promise<{
    inputs: Record<string, any>;
    outputs: Record<string, any>;
  }> {
    const response = await fetch(
      `${this.baseUrl}/api/agent-builder/block-types/${blockType}/schema`
    );
    if (!response.ok) {
      throw new Error(`Failed to fetch schema for block type: ${blockType}`);
    }
    return response.json();
  }

  /**
   * Get block registry statistics
   */
  async getStats(): Promise<BlockTypeStats> {
    const response = await fetch(
      `${this.baseUrl}/api/agent-builder/block-types/stats/summary`
    );
    if (!response.ok) {
      throw new Error('Failed to fetch block type stats');
    }
    return response.json();
  }

  /**
   * Get agentic block types (convenience method)
   */
  async getAgenticBlocks(): Promise<BlockTypeConfig[]> {
    return this.getBlockTypesByCategory('agentic');
  }
}

// Export singleton instance
export const blockTypesAPI = new BlockTypesAPI();

// Export for testing
export { BlockTypesAPI };

File: backend/api/agent_builder/block_types.py
"""
Block Types API

Provides endpoints for listing available block types from the BlockRegistry.
This includes built-in blocks and agentic workflow blocks.
"""

import logging
from typing import Dict, List, Any
from fastapi import APIRouter, Query

from backend.core.blocks import BlockRegistry

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/api/agent-builder/block-types",
    tags=["agent-builder-block-types"],
)


@router.get(
    "",
    response_model=Dict[str, List[Dict[str, Any]]],
    summary="List all available block types",
    description="Get all registered block types grouped by category"
)
async def list_block_types():
    """
    List all available block types from the BlockRegistry.
    
    Returns blocks grouped by category:
    - blocks: Standard workflow blocks
    - tools: Tool integration blocks
    - triggers: Event trigger blocks
    - agentic: Agentic workflow blocks (NEW)
    
    **Returns:**
    ```json
    {
        "agentic": [
            {
                "type": "agentic_rag",
                "name": "Agentic RAG",
                "description": "Intelligent retrieval with query decomposition...",
                "category": "agentic",
                "bg_color": "#EC4899",
                "icon": "brain",
                "sub_blocks": [...],
                "inputs": {...},
                "outputs": {...}
            },
            ...
        ],
        "blocks": [...],
        "tools": [...]
    }
    ```
    """
    try:
        blocks_by_category = BlockRegistry.list_by_category()
        logger.info(f"Listed {len(blocks_by_category)} block categories")
        return blocks_by_category
    except Exception as e:
        logger.error(f"Error listing block types: {e}", exc_info=True)
        return {}


@router.get(
    "/category/{category}",
    response_model=List[Dict[str, Any]],
    summary="List block types by category",
    description="Get all block types in a specific category"
)
async def list_block_types_by_category(category: str):
    """
    List block types filtered by category.
    
    **Parameters:**
    - category: Category name (agentic, blocks, tools, triggers)
    
    **Returns:**
    - List of block configurations in the specified category
    """
    try:
        blocks = BlockRegistry.list_blocks(category=category)
        logger.info(f"Listed {len(blocks)} blocks in category '{category}'")
        return blocks
    except Exception as e:
        logger.error(f"Error listing blocks for category '{category}': {e}", exc_info=True)
        return []


@router.get(
    "/{block_type}",
    response_model=Dict[str, Any],
    summary="Get block type configuration",
    description="Get detailed configuration for a specific block type"
)
async def get_block_type(block_type: str):
    """
    Get configuration for a specific block type.
    
    **Parameters:**
    - block_type: Block type identifier (e.g., "agentic_rag", "openai")
    
    **Returns:**
    - Block configuration with schema, sub_blocks, and metadata
    
    **Errors:**
    - 404: Block type not found
    """
    from fastapi import HTTPException
    
    config = BlockRegistry.get_block_config(block_type)
    if not config:
        raise HTTPException(
            status_code=404,
            detail=f"Block type '{block_type}' not found"
        )
    
    return config


@router.get(
    "/{block_type}/schema",
    response_model=Dict[str, Any],
    summary="Get block type schema",
    description="Get input/output schema for a specific block type"
)
async def get_block_type_schema(block_type: str):
    """
    Get JSON schema for block inputs and outputs.
    
    **Parameters:**
    - block_type: Block type identifier
    
    **Returns:**
    ```json
    {
        "inputs": {
            "query": {"type": "string", "required": true},
            "strategy": {"type": "string", "default": "adaptive"}
        },
        "outputs": {
            "answer": {"type": "string"},
            "confidence_score": {"type": "number"}
        }
    }
    ```
    
    **Errors:**
    - 404: Block type not found
    """
    from fastapi import HTTPException
    
    schema = BlockRegistry.get_block_schema(block_type)
    if not schema:
        raise HTTPException(
            status_code=404,
            detail=f"Block type '{block_type}' not found"
        )
    
    return schema


@router.get(
    "/stats/summary",
    response_model=Dict[str, Any],
    summary="Get block registry statistics",
    description="Get summary statistics about registered blocks"
)
async def get_block_stats():
    """
    Get statistics about the block registry.
    
    **Returns:**
    ```json
    {
        "total_blocks": 15,
        "by_category": {
            "agentic": 4,
            "blocks": 6,
            "tools": 3,
            "triggers": 2
        },
        "block_types": ["agentic_rag", "agentic_reflection", ...]
    }
    ```
    """
    try:
        blocks_by_category = BlockRegistry.list_by_category()
        block_types = BlockRegistry.get_block_types()
        
        stats = {
            "total_blocks": len(block_types),
            "by_category": {
                category: len(blocks)
                for category, blocks in blocks_by_category.items()
            },
            "block_types": block_types,
            "categories": list(blocks_by_category.keys())
        }
        
        return stats
    except Exception as e:
        logger.error(f"Error getting block stats: {e}", exc_info=True)
        return {
            "total_blocks": 0,
            "by_category": {},
            "block_types": [],
            "categories": []
        }
